### ISSUE TITLE
Setting locale, environment variables

### ISSUE THREAD

**Issue by ianclarksmith** [rogual/neovim-dot-app#54]
The issue I am experiencing is described here https://github.com/fish-shell/fish-shell/issues/277. This can be worked around by setting LC_ALL in one's fish config however it is "better" if the variables can be set with the terminal emulator. Example:

**`nvim` in iTerm**
```
fish> nvim
nvim> :terminal
nvim-terminal> locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```

**Neovim.app launched with `open`**
```
fish> open -a Neovim
Neovim.app> :terminal
Neovim.app-terminal> locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```

**Neovim.app launched from desktop**
```
Neovim.app> :terminal
Neovim.app-terminal> locale
LANG=
LC_COLLATE="C"
LC_CTYPE="C"
LC_MESSAGES="C"
LC_MONETARY="C"
LC_NUMERIC="C"
LC_TIME="C"
LC_ALL=
```

Emacs.app has a similar issue which led to the creation of the `exec-path-from-shell`. Without it, the `exec-path` in Emacs depends on the system-wide settings in a plist (depending on which OS X version).

Would it be better to automatically adopt the configuration of the user's login environment at launch, set environment variables manually (as TextMate does), or leave it entirely up each program's own configs? The first might be closer to the expected behavior while the last requires no changes.

**Comment by rogual** [comment]
Hi, thanks for the issue report. I'm not too familiar with Unix locales or Fish; from what I can gather, the issue is:

* Some terminal programs expect the LC_* vars to be set
* OSX doesn't have them set system-wide
* Terminal.app sets them on startup
* Vim doesn't

I'm going to continue assuming the above is correct, but please correct me if I'm wrong.

Firstly, is it the job of the terminal emulator to set LC_* vars? You say it's better if the terminal does this rather than your shell config. Why is this?

If the terminal emulator does need to set these vars, I wouldn't necessarily be against having Neovim.app set them, But, I'd say Neovim itself is what is emulating a terminal, so shouldn't that happen in there?

**Comment by ianclarksmith** [comment]
See here (http://apple.stackexchange.com/a/83433) for how Termial & iTerm do it. The default in Neovim.app is to use the "C" locale which is ASCII. OS X actually does have a system-wide locale but I don't think it is set by default for Cocoa applications.

The issue is: It is reasonable to expect neovim to perform the same whether launched from the terminal or GUI. MacVim does this automatically.

In the case of locale specifically, users should only have to set this once either with the system or their login shell and expect it to be picked up by all other programs. Cocoa apps generally don't need or want to take in the whole environment, but for dev tools it's helpful even if it isn't technically its responsibility.

In summary: There are many workarounds, but it makes everyone's configs simpler and reduces unexpected behavior if neovim in the GUI mirrors the environment of neovim in a terminal.

**Comment by rogual** [comment]
Would it solve your problem if the above linked pull request were merged?

**Comment by ianclarksmith** [comment]
Unfortunately not– it sets everything to "en_US" (not "en_US.UTF-8") and blows up once I start typing non-ASCII. By "blows up" I mean the cursor jumps to random places, changes color, and typing or attempting to exit don't work. I think this is the relevant function in iTerm which should help: https://github.com/gnachman/iTerm2/blob/11d9ea7d0f5625db9f42f1e2527c4a8c1e4fea5e/sources/PTYSession.m#L1019

Just noticed something else though: I can't input non-ASCII characters with the keyboard in Neovim.app even in a normal buffer. I can insert some characters through the keyboard viewer but no accented characters, kana, etc. Not a font issue (checking the obvious) and it's fine in the terminal so it appears to be an issue in Neovim.app (or maybe msgpack?).

[Apologies for not being more helpful with code– my Cocoa knowledge is screaming hello world with party streamers.]

**Comment by rogual** [comment]
Weird, I get "en_GB.UTF-8". Going to have to look into this one some more.

**Comment by kopischke** [comment]
OS X has a few <strike>infuriating</strike>fun quirks when it comes to setting the shell locale – one of them bit Neovim recenttly, maybe the discussion at https://github.com/neovim/neovim/issues/2386 can help.

**Comment by rogual** [comment]
Interesting stuff, thanks for the link. I've done a bit more work on that branch — @ianclarksmith, would you be able to try it out and see if it solves your issue?

**Comment by ianclarksmith** [comment]
@rogual Last commit on terminal-locale is from 3 days ago– push?

**Comment by rogual** [comment]
No, the date's just wrong because I did `--amend`. The branch is up to date.

**Comment by ianclarksmith** [comment]
Ah, alright. As far as I can tell, this solves the issue of my shell blowing up on certain character inputs. `LANG` is still not set though, which if I understand right would cause issues for some people regarding neovim/neovim#2386. This is apparently because Neovim or libtermkey doesn't follow POSIX: see also neovim/issues/2393 which evolved from that thread. Adding `setenv("LANG", ss.str().c_str(), 1);` sets it and has no adverse effects that I can detect.

However, it seems a more important issue is that many unicode characters can't be input or even displayed. For example, when I paste `日本語` I see `日本___` (`_` represents blank space). Typing with kotoeri input yields `nihongo` (actual keys pressed), while input from kana table or character viewer don't register anything. In a directory with an existing folder named `日本語` the ls command shows the same as when I attempt to paste (`日本___`). European accents I can paste and display but not type for some reason. All of this works fine in `nvim` launched from the shell though which is encouraging, but building with the `LANG` set didn't solve it.

I'm going to put the above in a separate issue since the narrow issue here of setting the locale can probably be closed.

**Comment by rogual** [comment]
PR merged (with LANG fix)
